## Importing the Libraries

In [1]:
import os

import numpy as np

import pandas as pd

from sklearn.model_selection import train_test_split

## READING THE DATA

In [2]:
# Here we use the r mean raw string to avoid any error

PROJECT_DIR = r'E:\FLIGHT PREDICTION PROJECT'
DATA_DIR = 'DATA'

In [3]:
# We craete a function to read any csv file just give the name of file the function give us the detail of data
# we need not to write read_csv again and again

def get_data(data_name):
    file_name = f'{data_name}.csv'
    file_path = os.path.join(PROJECT_DIR,DATA_DIR,file_name)
    read_file = pd.read_csv(file_path)
    return read_file

In [4]:
flights = get_data('flight_price')
flights

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [5]:
flights.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [6]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


## ABOUT DATA SET

- Data set have 10683 rows and 11 columns

- Column `Route` and Column `Total_Stops` have some missing values

- Some Columns have not appropriate data type

## PRELIMNARY DATA ANALYSIS

### Check Data Types

In [7]:
flights.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [8]:
# to check the data types of each feature
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

- Columns (`Date_of_Journey,Dep_Time,Arrival_Time,Duration`) heve not a suitable data type so we need to change the data type of these columns in appropriate data type


## CHECK FOR DUPLICATES

In [9]:
# We use pandas dulicated function to check duplicated values in our data set
flights.duplicated().sum()

220

In [10]:
# Now we extract the duplicated rows from our data set
# For extracting the duplicated rows we perform the boolean indexing on data set
# in this code we keep false in duplicated function because we want to show the all duplictaed values
flights[flights.duplicated(keep=False)]


Airline Date_of_Journey    Source Destination  \
33     Jet Airways      15/06/2019     Delhi      Cochin   
49     Jet Airways      27/06/2019     Delhi      Cochin   
73     Jet Airways      24/06/2019     Delhi      Cochin   
81     Jet Airways      24/03/2019  Banglore   New Delhi   
87     Jet Airways       6/06/2019     Delhi      Cochin   
...            ...             ...       ...         ...   
10594  Jet Airways      27/06/2019     Delhi      Cochin   
10616  Jet Airways       1/06/2019     Delhi      Cochin   
10634  Jet Airways       6/06/2019     Delhi      Cochin   
10672  Jet Airways      27/06/2019     Delhi      Cochin   
10673  Jet Airways      27/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
33     DEL → NAG → BOM → COK    14:35  12:35 16 Jun      22h     2 stops   
49     DEL → AMD → BOM → COK    23:05  19:00 28 Jun  19h 55m     2 stops   
73     DEL → AMD → BOM → COK    23:05  12:35 25 Jun  13h 30m     2 stops   
81                 BLR → DEL    19:55         22:35   2h 40m    non-stop   
87     DEL → JAI → BOM → COK    09:40  04:25 07 Jun  18h 45m     2 stops   
...                      ...      ...           ...      ...         ...   
10594  DEL → AMD → BOM → COK    23:05  12:35 28 Jun  13h 30m     2 stops   
10616  DEL → JAI → BOM → COK    09:40  12:35 02 Jun  26h 55m     2 stops   
10634  DEL → JAI → BOM → COK    09:40  12:35 07 Jun  26h 55m     2 stops   
10672  DEL → AMD → BOM → COK    23:05  19:00 28 Jun  19h 55m     2 stops   
10673  DEL → AMD → BOM → COK    13:25  04:25 28 May      15h     2 stops   

                   Additional_Info  Price  
33     In-flight meal not included  10919  
49     In-flight meal not included  11150  
73                         No info  12819  
81                         No info   7229  
87                         No info  13014  
...                            ...    ...  
10594                      No info  12819  
10616                      No info  13014  
10634  In-flight meal not included  11733  
10672  In-flight meal not included  11150  
10673                      No info  16704  

[415 rows x 11 columns]

In [11]:
# we use another method to check the all duplicates
(
    flights
    .loc[flights.duplicated(keep=False)]
#     we use sort values function to sort all values because we check see all duplicated values in sequence
    .sort_values(['Airline','Date_of_Journey','Source','Destination'])
    
)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

In [12]:
# with out using keep False
flights[flights.duplicated()]

Airline Date_of_Journey    Source Destination  \
683    Jet Airways       1/06/2019     Delhi      Cochin   
1061     Air India      21/05/2019     Delhi      Cochin   
1348     Air India      18/05/2019     Delhi      Cochin   
1418   Jet Airways       6/06/2019     Delhi      Cochin   
1674        IndiGo      24/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10594  Jet Airways      27/06/2019     Delhi      Cochin   
10616  Jet Airways       1/06/2019     Delhi      Cochin   
10634  Jet Airways       6/06/2019     Delhi      Cochin   
10672  Jet Airways      27/06/2019     Delhi      Cochin   
10673  Jet Airways      27/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
683    DEL → NAG → BOM → COK    14:35  04:25 02 Jun  13h 50m     2 stops   
1061   DEL → GOI → BOM → COK    22:00  19:15 22 May  21h 15m     2 stops   
1348   DEL → HYD → BOM → COK    17:15  19:15 19 May      26h     2 stops   
1418   DEL → JAI → BOM → COK    05:30  04:25 07 Jun  22h 55m     2 stops   
1674               BLR → DEL    18:25         21:20   2h 55m    non-stop   
...                      ...      ...           ...      ...         ...   
10594  DEL → AMD → BOM → COK    23:05  12:35 28 Jun  13h 30m     2 stops   
10616  DEL → JAI → BOM → COK    09:40  12:35 02 Jun  26h 55m     2 stops   
10634  DEL → JAI → BOM → COK    09:40  12:35 07 Jun  26h 55m     2 stops   
10672  DEL → AMD → BOM → COK    23:05  19:00 28 Jun  19h 55m     2 stops   
10673  DEL → AMD → BOM → COK    13:25  04:25 28 May      15h     2 stops   

                   Additional_Info  Price  
683                        No info  13376  
1061                       No info  10231  
1348                       No info  12392  
1418   In-flight meal not included  10368  
1674                       No info   7303  
...                            ...    ...  
10594                      No info  12819  
10616                      No info  13014  
10634  In-flight meal not included  11733  
10672  In-flight meal not included  11150  
10673                      No info  16704  

[220 rows x 11 columns]

## OBSETVATIONS ABOUT DUPLICATED VALUES

- Columns `Date_of_Journey`,`Dep_Time`,`Arrival_Time` should be converted into date_time data type

- Columns `Duration` and `Total_Stops` is mixed it shuold be converted into numeric data type

- in this data 220 values are duplicated it should be removed

## DETAILED DATA ANALYSIS

### UNIVARIATE ANALYSIS

## Airline 

In [13]:
# Now we perform colum by column analysis
# 1st column Airline
# we access the Airline column from dataset

flights['Airline']


0             IndiGo
1          Air India
2        Jet Airways
3             IndiGo
4             IndiGo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

In [14]:
# To check the number of unique values
flights.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

## OBSERVATIONS

- Some inconsistent value in Airline column like `Vistara Premium economy` and `Jet Airways Business` so we correct it

In [15]:
# This is an alternative method to change the column but we use assign method of pandas in function
# flights['airline'] = flights.Airline.str.replace('Premium economy','').str.replace('Business','').str.title()


In [16]:
# This is just testing code beacuse we do all things in function
# To Increas the code readability we also write code in this manner 
# so in our whole project i try to use this method for writing code
# in this step i remove the inconsistent value like Business and Premium economy
(
    flights
    .Airline
#     we replace the incosistent values 
    .str.replace('Premium economy','')
    .str.replace('Business','')
#     we use title method to capital the first leeter of each word
    .str.title()




)

0             Indigo
1          Air India
2        Jet Airways
3             Indigo
4             Indigo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

In [17]:
flights.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

## Date_of_Journey

`Data Type of this column is inccoreect we convert it data type in date and time format`

In [18]:
# This is a testing code
# we do all thing in function
pd.to_datetime(flights.Date_of_Journey,dayfirst=True)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

## Source

In [19]:
flights.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

In [20]:
flights.Source.value_counts()

Delhi       4537
Kolkata     2871
Banglore    2197
Mumbai       697
Chennai      381
Name: Source, dtype: int64

- `We have 5 different sources `

## Destination

In [21]:
flights.Destination.value_counts()

Cochin       4537
Banglore     2871
Delhi        1265
New Delhi     932
Hyderabad     697
Kolkata       381
Name: Destination, dtype: int64

In [22]:
flights.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

In [23]:
flights.Destination.nunique()

6

- `We have 6 differnet destinations in this data`

## Route

In [24]:
# testting code
# we remove this column because this column contain redundent information
# so we remove this column
flights.Route

0                    BLR → DEL
1        CCU → IXR → BBI → BLR
2        DEL → LKO → BOM → COK
3              CCU → NAG → BLR
4              BLR → NAG → DEL
                 ...          
10678                CCU → BLR
10679                CCU → BLR
10680                BLR → DEL
10681                BLR → DEL
10682    DEL → GOI → BOM → COK
Name: Route, Length: 10683, dtype: object

- `in this column we have a mixed data so we need to correct this column`
- `May be this column is not much useful so we delete this coloumn in future`

## Dep_Time

In [25]:
# We need to covert the data type into date time
flights.Dep_Time

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

In [26]:
flights.Dep_Time.loc[lambda ser:ser.str.contains('[^0-9]:')]

Series([], Name: Dep_Time, dtype: object)

In [27]:
# Now we convert the Dep_Time column into date and time data type
pd.to_datetime(flights.Dep_Time).dt.time

0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

## Arrival_Time

- `This column contain mixed data with time and month name`

- `we need to separate the time information and month information`

In [28]:
(
    flights
    .Arrival_Time
    .loc[lambda ser : ser.str.contains('[^0-9:]')]
    .str.split(" ",1)
    .str.get(1)
    .unique()
    
)

array(['22 Mar', '10 Jun', '13 Mar', '02 Mar', '10 May', '04 Mar',
       '13 Jun', '28 May', '19 Mar', '07 May', '02 Jun', '16 Jun',
       '19 May', '16 May', '28 Jun', '02 May', '28 Mar', '19 Jun',
       '04 Apr', '25 Mar', '07 Mar', '25 Jun', '07 Jun', '25 May',
       '13 May', '16 Mar', '22 May', '10 Apr', '04 Jun', '20 May',
       '28 Apr', '25 Apr', '10 Mar', '19 Apr', '13 Apr', '02 Apr',
       '23 Mar', '22 Apr', '11 May', '07 Apr', '03 May', '08 Mar',
       '03 Mar', '05 Mar', '22 Jun', '04 May', '26 May', '16 Apr',
       '26 Jun', '29 May', '29 Jun', '29 Mar', '23 May', '17 Jun'],
      dtype=object)

In [29]:
# Altrenative code
flights.Arrival_Time.str.extract(r'(\d{2}:\d{2})')

# (
#     flights
#     .Arrival_Time
#     .loc[lambda ser : ser.str.contains('[^0-9:]')]
#     .str.split(' ',1)
#     .str.get(1)
#     .unique()
    

# )


0
0      01:10
1      13:15
2      04:25
3      23:30
4      21:35
...      ...
10678  22:25
10679  23:20
10680  11:20
10681  14:10
10682  19:15

[10683 rows x 1 columns]

## Duration

In [30]:
flights.Duration.loc[lambda ser : ~ser.str.contains('m')].unique()

array(['19h', '23h', '22h', '12h', '3h', '5h', '10h', '18h', '24h', '15h',
       '16h', '8h', '14h', '20h', '13h', '11h', '9h', '27h', '26h', '4h',
       '7h', '30h', '21h', '28h', '47h', '6h', '25h', '38h', '34h'],
      dtype=object)

In [31]:
flights.Duration.loc[lambda ser : ~ser.str.contains('h')]

6474    5m
Name: Duration, dtype: object

In [32]:
flights.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

`In row# 6474 the flight duration is only 5 minute this is incorrect so we delete this row`

In [33]:
(
    flights
    .Duration
    .drop(index=6474)
#     we use expand method to convert it into datafraem 
    .str.split(' ',expand=True)
#     we use set_axis for giving name to columns
    .set_axis(['hours','minutes'],axis=1)
    .assign(
        hours = lambda df_ :(
            df_
            .hours
            .str.replace('h','')
            .astype(int)
            .mul(60)
        ),
        minutes = lambda df_ : (
            df_
            .minutes
            .str.replace('m','')
            .fillna('0')
            .astype(int)
            
        )
    )
    .sum(axis=1)
#     we rename this series
    .rename('duration_minutes')
#     we convert this series into dataframe
    .to_frame()
#     we join with original data frame for validating the values
    .join(flights.Duration.drop(index=6474))

)

duration_minutes Duration
0                   170   2h 50m
1                   445   7h 25m
2                  1140      19h
3                   325   5h 25m
4                   285   4h 45m
...                 ...      ...
10678               150   2h 30m
10679               155   2h 35m
10680               180       3h
10681               160   2h 40m
10682               500   8h 20m

[10682 rows x 2 columns]

In [34]:
# Now we convert the above block of code into function because we want to add the time information into original data frame
def convert_to_minutes(duration):
    
    return (
        duration#.drop(index=6474)
        .str.split(' ',expand=True)
        .set_axis(['hour','minute'],axis=1)
        .assign(
            hour = lambda df_:df_.hour.str.replace('h','').astype(int).mul(60),
            minute = lambda df_:df_.minute.str.replace('m','').fillna('0').astype(int)
        )
        .sum(axis=1)
    )
    
# convert_to_minutes(flights.Duration)  
# flights.Duration.pipe(convert_to_minutes)

In [35]:
# This is an alternative function to convert the hour and minutes information into total minutes
# we convert the the hours and minutes informations into total minutes
# flights.Duration

# def convert_to_minutes(duration):
#     hours = 0
#     minutes = 0
#     if 'h' in duration:
#         hours = int(duration.split('h')[0])
#         duration = duration.split('h')[1]
#     if 'm' in duration:
#         minutes = int(duration.split('m')[0])
                      
#     total_minutes = hours*60 + minutes
#     return total_minutes
                      

# flights.Duration.apply(convert_to_minutes)   

## Total Stops

In [36]:
flights.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [37]:
(
    flights
    .Total_Stops
    .replace('non-stop','0')
    .str.replace('stops?','',regex=True)
#     we cannot convert the data type as integer bacause we have some missing values
#     .astype(int)
#     so we use pandas pipe function to convert data into numeric values
#     we use lmbda function to get the current state of series
    .pipe(lambda ser : pd.to_numeric(ser))
    
    
)

0        0.0
1        2.0
2        2.0
3        1.0
4        1.0
        ... 
10678    0.0
10679    0.0
10680    0.0
10681    0.0
10682    2.0
Name: Total_Stops, Length: 10683, dtype: float64

## Additional Info

In [38]:
flights.Additional_Info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

In [39]:
flights.Additional_Info.replace('No info','No Info')

0        No Info
1        No Info
2        No Info
3        No Info
4        No Info
          ...   
10678    No Info
10679    No Info
10680    No Info
10681    No Info
10682    No Info
Name: Additional_Info, Length: 10683, dtype: object

## Cleaning data operations

In [48]:
# we create a function for data cleaning because we use functional programming to perfrom data cleaning operatins
# so we create function for performing data cleaning operation this functio receive data frame
# we use method chaining rule because we want to avoid the changes in original data
# we can chage our data but pandas does not recomend it .
# changing the original data is not good approach

def cleaned_data(df):
#     this function return a cleaned data frame
    return(
        df
#         we drop the row no 6474 because it cotain wrong duratin it has just only 5m of flight duration
#         so we drop this row
        .drop(index=[6474])
#         we remove the duplicated rows
        .drop_duplicates()
#          we use assign method to create new column or updating the existing columns
#         so we use assign method because we want to update the columns in data frame
#         we write the column name want to change or update
#         airline = df.Ailine.str.lower() we use this code but we use lambda function because automatically call the data
#         This function lower the all values of airlie column
#         pandas will automatically pass the current state of the data frame not the original one
#         airline = lambda df1:df1.airline.str.lower() this code is use for sigle column that has sting values
#         we also use this code but we use lambda function because this function automatically call the data
#         airline = df.Ailine.str.lower()
#         we rename the all columns in lower case because it is good approach
        .assign(**{
#             we use dictionary comprehension concept here
            col : df[col].str.strip()
            for col in df.select_dtypes(include='O').columns
            
        })
            
#        in this step we rename the all columns into lowercase     
        .rename(columns=str.lower)
#         in this step we perform the operation on Airline column and remove the inconsistent data
        .assign(
#             we use the lambda function to getting the current state of the dataframe
            airline = lambda df_:(
                
                df_
#                 Airline column
                .airline
#             these two line of code replace the incosistent values 
                .str.replace('Premium economy','')
                .str.replace('Business','')
#             we use title method to capital the first leeter of each word
                .str.title()
            
            ),
#             Date of journy column
#             Now convert the data type of date_of_journey column into data time
            date_of_journey = lambda df_:pd.to_datetime(df_.date_of_journey,dayfirst = True),
#             Dep_Time column
#             Now convert the data type of dep_time column into data time
            dep_time = lambda df_:pd.to_datetime(df_.dep_time).dt.time,
#             Arrival_Time column
#             Now convert the data type of arrival_time column into data time
            arrival_time = lambda df_:pd.to_datetime(df_.arrival_time).dt.time,
#             Duration column
#             Now we add the convert_to_minutes function for adding the total tim information in original dataframe
#             add the convert_to_minutes function by using pipe method of pandas
#             we use this code but we use pandas pipe  method
#             duration = lambda df_ : convert_to_minutes(df_.duration)
#             we use pandas pipe method
            duration = lambda df_ : df_.duration.pipe(convert_to_minutes),
#             Total_Stops column
#             we convert this column into numeric form
            total_stops = lambda df_:(
                df_
                .total_stops
                .replace('non-stop','0')
                .str.replace('stops?','',regex=True)
                .pipe(lambda ser : pd.to_numeric(ser))
            ),
            
            additional_info = lambda df_ :(
                df_
                .additional_info
                .replace('No info','No Info')
            )
            
                
                    
        )
#         route column
#         removing the route column because it consist on redundent information
        .drop(columns='route')
#        
        
    )

In [49]:
flights_cleaned = cleaned_data(flights)
flights_cleaned

airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No Info   3897  
1         13:15:00       445          2.0         No Info   7662  
2         04:25:00      1140          2.0         No Info  13882  
3         23:30:00       325          1.0         No Info   6218  
4         21:35:00       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No Info   4107  
10679     23:20:00       155          0.0         No Info   4145  
10680     11:20:00       180          0.0         No Info   7229  
10681     14:10:00       160          0.0         No Info  12648  
10682     19:15:00       500          2.0         No Info  11753  

[10462 rows x 10 columns]

## Train Test Split

In [53]:
x = flights_cleaned.drop(columns = 'price')
y = flights_cleaned.price

In [57]:
x_,x_test,y_,y_test = train_test_split(x,y,test_size = 0.2,random_state = 42)
x_train,x_val,y_train,y_val = train_test_split(x_,y_,test_size = 0.2,random_state = 42)

In [68]:
print('Training Data')
print(x_train.shape,y_train.shape)
print('Validation Data')
print(x_val.shape,y_val.shape)
print('Testing Data')
print(x_test.shape,y_test.shape)

Training Data
(6695, 9) (6695,)
Validation Data
(1674, 9) (1674,)
Testing Data
(2093, 9) (2093,)


In [60]:
def export_data(x,y,name):
    file_name = f"{name}.csv"
    file_path = os.path.join(PROJECT_DIR,DATA_DIR,file_name)
    x.join(y).to_csv(file_path,index = False)
    return pd.read_csv(file_path).head()

In [61]:
export_data(x_train,y_train,'train_data')

airline date_of_journey    source destination  dep_time arrival_time  \
0  Jet Airways      2019-03-21  Banglore   New Delhi  08:55:00     19:10:00   
1  Jet Airways      2019-03-27     Delhi      Cochin  17:30:00     04:25:00   
2        Goair      2019-03-09  Banglore   New Delhi  11:40:00     14:35:00   
3    Air India      2019-06-12   Kolkata    Banglore  09:25:00     18:30:00   
4  Jet Airways      2019-03-12  Banglore   New Delhi  22:55:00     07:40:00   

   duration  total_stops              additional_info  price  
0       615          1.0  In-flight meal not included   7832  
1       655          1.0  In-flight meal not included   6540  
2       175          0.0                      No Info   7305  
3       545          1.0                      No Info   8366  
4       525          1.0  In-flight meal not included  11087

In [62]:
export_data(x_test,y_test,'test_data')

airline date_of_journey    source destination  dep_time arrival_time  \
0  Jet Airways      2019-03-06  Banglore   New Delhi  08:00:00     08:15:00   
1     Spicejet      2019-06-06   Kolkata    Banglore  22:20:00     00:40:00   
2       Indigo      2019-03-18   Kolkata    Banglore  05:30:00     08:20:00   
3       Indigo      2019-06-27   Chennai     Kolkata  19:35:00     21:55:00   
4       Indigo      2019-05-06   Kolkata    Banglore  15:15:00     17:45:00   

   duration  total_stops additional_info  price  
0      1455          1.0         No Info  17996  
1       140          0.0         No Info   3873  
2       170          0.0         No Info   4462  
3       140          0.0         No Info   3597  
4       150          0.0         No Info   4804

In [63]:
export_data(x_val,y_val,'validation_data')

airline date_of_journey    source destination  dep_time  \
0             Indigo      2019-06-24     Delhi      Cochin  20:25:00   
1  Multiple Carriers      2019-06-12     Delhi      Cochin  09:45:00   
2        Jet Airways      2019-03-12  Banglore   New Delhi  22:55:00   
3  Multiple Carriers      2019-06-06     Delhi      Cochin  13:00:00   
4        Jet Airways      2019-05-18     Delhi      Cochin  23:05:00   

  arrival_time  duration  total_stops              additional_info  price  
0     01:30:00       305          1.0                      No Info   5054  
1     22:30:00       765          1.0                      No Info   9646  
2     15:15:00       980          1.0  In-flight meal not included  11087  
3     21:00:00       480          1.0                      No Info  13587  
4     04:25:00      1760          2.0                      No Info  16704

In [54]:
# now test the data change are amade or not
cleaned_data(flights).airline.unique()

array(['Indigo', 'Air India', 'Jet Airways', 'Spicejet',
       'Multiple Carriers', 'Goair', 'Vistara', 'Air Asia', 'Vistara ',
       'Jet Airways ', 'Multiple Carriers ', 'Trujet'], dtype=object)

In [43]:
# now test the date_of_journey column changes
cleaned_data(flights).dtypes

airline                    object
date_of_journey    datetime64[ns]
source                     object
destination                object
dep_time                   object
arrival_time               object
duration                    int64
total_stops               float64
additional_info            object
price                       int64
dtype: object

In [44]:
# Now test the dep_time column
cleaned_data(flights).dtypes

airline                    object
date_of_journey    datetime64[ns]
source                     object
destination                object
dep_time                   object
arrival_time               object
duration                    int64
total_stops               float64
additional_info            object
price                       int64
dtype: object

In [45]:
cleaned_data(flights).dtypes

airline                    object
date_of_journey    datetime64[ns]
source                     object
destination                object
dep_time                   object
arrival_time               object
duration                    int64
total_stops               float64
additional_info            object
price                       int64
dtype: object

In [46]:
# Alternative code for renaming the columns into lower case
# new = []
# for x in flights:
# #     print(x)
#     new_col= x.lower()
#     new.append(new_col)
# new
flights.select_dtypes(include = 'O').columns 

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info'],
      dtype='object')

In [47]:
# this code verify
flights['Dep_Time'].str.contains('[^0-9:]').sum()


0